### PR TITLE
Write model metadata to export folder unconditionally

### PR DESCRIPTION
**Pull request by reuben**

### CODE REVIEW

**Comment by reuben** [comment]
Example output:

```
---
author: author
model_name: model
model_version: 1
contact_info: <public contact information of the author. Free form, could be an email address, a repository URL, a company website, etc.>
license: <license of the exported model>
language: <language the model was trained on - IETF BCP 47 language tag, e.g. "de-DE" or "zh-cmn-Hans-CN">
runtime: tensorflow
min_ds_version: <minimum DeepSpeech version (inclusive) the exported model is compatible with>
max_ds_version: <maximum DeepSpeech version (inclusive) the exported model is compatible with>
acoustic_model_url: <replace this with a publicly available URL of the acoustic model>
scorer_url: <replace this with a publicly available URL of the scorer, if present>
---
<Freeform description of the model being exported. Markdown accepted. You can also leave this flag unchanged and edit the generated .md file directly. Useful things to describe are demographic and acoustic characteristics of the data used to train the model, any architectural changes, names of public datasets that were used when applicable, hyperparameters used for training, evaluation results on standard benchmark datasets, etc.>
```

**Comment by lissyx** [comment]
Wouldn't we want that kind of metadata to be easily parsable by a machine? And thus use a higher-level description than just `markdown` ?

**Comment by kdavis-mozilla** [comment]
@lissyx The idea is to use Jekyll [collections](https://jekyllrb.com/docs/collections/). (The file generated by the PR's code would be the analog of the `jane.md` file in the example on the linked page.)

**Comment by lissyx** [comment]
> @lissyx The idea is to use Jekyll [collections](https://jekyllrb.com/docs/collections/). (The file generated by the PR's code would be the analog of the `jane.md` file in the example on the linked page.)

What would be the use-case here ?

**Comment by kdavis-mozilla** [comment]
We're going to have a Mozilla ML website, a Jekyll GitHub page, and as part of that we will have a model repository where internal and external models can be advertised. Adding a model will essentially be submitting a PR with the .md generated by this code.

File: util/flags.py [hunk 1]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
     f.DEFINE_integer('n_steps', 16, 'how many timesteps to process at once by the export graph, higher values mean more latency')
-    f.DEFINE_string('export_language', '', 'language the model was trained on e.g. "en" or "English". Gets embedded into exported model.')
     f.DEFINE_boolean('export_zip', False, 'export a TFLite model and package with LM and info.json')
-    f.DEFINE_string('export_name', 'output_graph', 'name for the export model')
+    f.DEFINE_string('export_file_name', 'output_graph', 'name for the exported model file name')
     f.DEFINE_integer('export_beam_width', 500, 'default beam width to embed into exported graph')
 
+    # Model metadata
+
+    f.DEFINE_string('export_author_id', 'author', 'author of the exported model. This is a unique handle used for identifying and collating models per author. Suggestion: use your GitHub username or organization name')
```
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: Also as this is used in the output file name there are likely constraints on the possible name used here, e.g. no forward slashes.

File: util/flags.py [hunk 2]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
+    f.DEFINE_string('export_model_name', 'model', 'name of the exported model')
```
Comment (kdavis-mozilla) [inline util/flags.py]: As this is used in the output file name there are likely constraints on the possible name used here, e.g. no forward slashes.

File: util/flags.py [hunk 3]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
+    f.DEFINE_string('export_model_version', '1', 'version of the exported model. This is fully controlled by you as author of the model and has no required connection with DeepSpeech versions')
```
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: Also as this is used in the output file name there are likely constraints on the possible name used here, e.g. no forward slashes.

File: util/flags.py [hunk 4]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
+
+    def str_val_equals_help(name, val_desc):
+        f.DEFINE_string(name, '<{}>'.format(val_desc), val_desc)
+
+    str_val_equals_help('export_contact_info', 'public contact information of the author. Free form, could be an email address, a repository URL, a company website, etc.')
```
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: Email address seems suboptimal, I'd say most open source maintainers would prefer to link to a repository where people can file issues, or a forum where people can create threads with questions.

File: util/flags.py [hunk 5]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
+    str_val_equals_help('export_license', 'license of the exported model')
```
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: No complete standard can exist, as anyone can make a new license if they so desire. But SPDX offers [identifiers for most common licenses](https://spdx.org/licenses/) that we can instruct people to use and maybe add some linting check to require more attention if a license is not in that list. Do we want to enforce any licensing limits? OSI approved only?

File: util/flags.py [hunk 6]
```diff
@@ -112,11 +112,26 @@ def create_flags():
...
+    str_val_equals_help('export_language', 'language the model was trained on - IETF BCP 47 language tag, e.g. "de-DE" or "zh-cmn-Hans-CN"')
```
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: If we add constraints we will limit what information can be conveyed here and will be making our own little derived standard. I'd rather just stick to the BCP if possible.

File: util/flags.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: I'll change it to be a semantic version.

File: util/flags.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: We would have to enforce no collisions anyway on our side. Enforcing GitHub user or organization name is probably the easiest to enforce, can even do programmatic linter checks, but I feel a bit weird about linking this requirement to an unrelated commercial service just for convenience.

File: util/flags.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: But leaving it just "BCP". Doesn't provide enough information.

For example I could write zh and be BCP compliant.

But zh doesn't say what script, e.g. Simplified Chinese or Traditional Chinese I've used. So someone will not know if they want to download it.

[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: Our pretrained models, for example, are probably best described as "en".

File: util/flags.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: I don't want to enforce OSI approved license, but I don't want N versions of the same license. So the SPDX strings could be the "standard" unless someone really wants to add a exception.

File: util/flags.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: Or en-US

File: util/flags.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: I think most models likely won't have a corresponding repository or forum as they're just a result of someone running a training run.

However, all models will have an associated GitHub user, as some has to do the PR. So maybe it should just be the GitHub user doing the PR?

[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: So should I use your user/email for our models? ;)

File: util/flags.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: SPDX it is then.

File: util/flags.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: Yeah, point taken :)

However, I think if something is amiss with a model, e.g. it doesn't perform as advertised or, more sinisterly, it exploits some unknown TF bug to take over your machine, there should be someone to contact.

**Comment by dabinat** [comment]
Perhaps it’s better to let users put whatever they want and do the validation on the server end when submitting a model to the store? This feature may be useful for adding metadata even for people not using the model store, so it would be great to not impose too many strict store-specific restrictions. 

So maybe it’s better to have the help text suggest sensible data formats but not aggressively validate them.

**Comment by kdavis-mozilla** [comment]
> Perhaps it’s better to let users put whatever they want...

Are you referring to a specific field or are you making the suggestion that the meta data should be completely unstructured text or are you making the suggestion that each existing field should not be validated?

File: util/flags.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: I'll enforce script and region subtags. Our English models would then become en-Latn-US and hopefully in the future en-Latn-001.

**Comment by reuben** [comment]
All of the validations we're talking about will be done server side. The discussion here is about the guidelines in the help text and default file that's generated.

File: util/flags.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: I changed the text, how do you feel about the new phrasing?

[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: I changed the text, how do you feel about the new phrasing?

[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: @kdavis-mozilla ping re the wording here. I updated it to "public contact information of the author. Can be an email address, or a link to a contact form, issue tracker, or discussion forum. Must provide a way to reach the model authors"

[the remark this answers is not in the document]
Comment (kdavis-mozilla) [reply util/flags.py]: It's alright if we assume the person supplying the model has the best intentions. But I don't know if we should assume the best intentions.

However, whatever the person places here, we will at worst have the contact info of the person making the PR.

[the remark this answers is not in the document]
Comment (reuben) [reply util/flags.py]: We can also reserve the right to remove models if we get no timely response.

**Pull request merged**